Doc/server types (#622)
pyansys/pydpf-core: docs/source/user_guide/server_types.rst

.. _user_guide_server_types:

===========================
Client-server communication
===========================

DPF is based on a client-server architecture.

> The DPF Server acts as a server, while the different client APIs (CPython, IronPython, C++...)

Reviewer: C++, and so on)

act as clients connecting to it.

> PyDPF is for example a term encompassing both the CPython client

Reviewer: PyDPF is a term encompassing both....

(the Python packages such as ansys-dpf-core or ansys-dpf-post, available on PyPI)
and the IronPython client (available within Ansys Mechanical).

The communication logic with a DPF server is defined when starting it using

> an instance of the :class:`ansys.dpf.core.server_factory.ServerConfig` class.

Author: ```suggestion
an instance of the :class:`ServerConfig<ansys.dpf.core.server_factory.ServerConfig>` class.
```

Different predefined server configurations are available in DPF,

> to better answer different use-cases

Reviewer 2: I don't understand "better" here

Author: @cbellot000 "better" might the wrong word. What I meant is "adapted"

Author: ```suggestion
each answering a different use-case
```

> (See the :class:`ansys.dpf.core.server_factory.AvailableServerConfigs` class).

Author: ```suggestion
(See the :class:`AvailableServerConfigs<ansys.dpf.core.server_factory.AvailableServerConfigs>` class).
```

> - The `GrpcServer` configuration is available starting with server version 4.0 (Ansys 2022R2).

Author: ```suggestion
- The :class:`GrpcServer <ansys.dpf.core.server_types>` configuration is available starting with server version 4.0 (Ansys 2022R2).
```

> - The :class:`GrpcServer <ansys.dpf.core.server_types>` configuration is available starting with server version 4.0 (Ansys 2022R2).

Reviewer: 2022 R2

  It allows for remote connections to a DPF Server across a network by telling the client
  to communicate with this server via the gRPC communication protocol.
  Although it can be used to communicate with a DPF server running on the same local machine,

>   in that case the next configuration should be preferred.

Reviewer: the next configuration is better for this option.

> - The `InProcess` configuration is available starting with server version 4.0 (Ansys 2022R2).

Author: ```suggestion
- The :class:`InProcess <ansys.dpf.core.server_types>` configuration is available starting with server version 4.0 (Ansys 2022R2).
```

> - The :class:`InProcess <ansys.dpf.core.server_types>` configuration is available starting with server version 4.0 (Ansys 2022R2).

Reviewer: 2022 R2

>   It tells the client that a DPF server is installed on the local machine, enabling direct calls

Reviewer 2: tell? maybe indicates?

Author: ```suggestion
  It indicates to the client that a DPF server is installed on the local machine, enabling direct calls
```

  to the server binaries from within the client's own Python process.
  This removes the need to copy and send data between the client and server, and makes calls
  to the server functionalities much faster as well as using less memory.

> - The `LegacyGrpcServer` configuration is the only one available for server versions below 4.0

Reviewer 2: we could maybe have links to the API page at each mention of LegacyGrpcServer, GrpcServer, InProcess

Author: @cbellot000 Sure. Will do.

Author: ```suggestion
- The :class:`LegacyGrpcServer<ansys.dpf.core.server_types.LegacyGrpcServer>` configuration is the only one available for server versions below 4.0
```

>   (Ansys 2022R1, Ansys 2021R2 and Ansys 2021R1).

Reviewer: 2022 R1, 2021 R2, 2021 R1

  The client communicates with a local or remote DPF server via the gRPC communication protocol.

> For DPF with Ansys 2023 R1 and newer, the default configuration is set to `InProcess`,

Author: ```suggestion
For DPF with Ansys 2023 R1 and newer, the default configuration is set to :class:`InProcess <ansys.dpf.core.server_types>`,
```

meaning that servers are launched on the local machine.
To launch a DPF server on a remote machine and communicate with it using gRPC, one should use

> the `GrpcServer` configuration as shown in :ref:`_ref_server_types_example`.

Author: ```suggestion
the :class:`GrpcServer <ansys.dpf.core.server_types>` configuration as shown in :ref:`_ref_server_types_example`.
```

> To launch a DPF server on a remote machine and communicate with it using gRPC, one should use

Reviewer: using gRPC, use the....

> the :class:`GrpcServer <ansys.dpf.core.server_types>` configuration as shown in :ref:`_ref_server_types_example`.

Reviewer 3: @PProfizi would it be possible to have code examples on how to use the server configurations?
